WordPress/gutenberg
# Style Engine: block supports backend

Tracking Issue: https://github.com/WordPress/gutenberg/issues/38167

## What?

A backend version of
- https://github.com/WordPress/gutenberg/pull/37978

~Experimenting with classname obfuscation.~ Reverted (see c[ommit history](https://github.com/WordPress/gutenberg/pull/39446/commits/eba46f468d8d3feeadca16ae2226c7d4872a0b6c#diff-df6c4ae64dee8330db2c09f5c60668cea91dfcbbad4e5143055c5c98e29ccbf5R106))

Works on block support styles (margin/padding for now) that are serverside rendered, e.g., `core/site-title` 

## Why?

To create a backend engine for style object -> css rule transformation to complement the frontend code: https://github.com/WordPress/gutenberg/pull/37978

## How?

`WP_Style_Engine_Gutenberg` has a `STYLE_DEFINITIONS_METADATA` dictionary that contains instructions on how to parse a Gutenberg style attributes in order to return CSS rules.

## Testing Instructions
Insert a few `core/site-title` blocks and add padding.

<details><summary>Editor code</summary>

 ```html







```
</details>


Check the resulting output and CSS inline styles.

```html
<h1 style="background-color: #ee8c8c; padding-top: 104px; padding-right: 104px; padding-bottom: 104px; padding-left: 104px;  margin-top: 286px; margin-bottom: 286px; margin-right: 286px; margin-left: 286px; " class="has-background wp-block-site-title"><a href="http://localhost:4759" rel="home">a8c-wp-env</a></h1>
```
## Related PRs

- https://github.com/WordPress/gutenberg/pull/37978
- https://github.com/WordPress/gutenberg/pull/38974
- https://github.com/WordPress/gutenberg/pull/39374
